server/src/inmantals/server.py

```diff
@@ -190,84 +228,84 @@ def create_tmp_project(self) -> str:
-        """Reads the default value for a setting set by the initializationOptions argument in the initialize method."""
-        logger.info(f"key={key} self.inmanta_settings={self.inmanta_settings}")
-        return self.inmanta_settings.get(key, default)
-
-    def remove_folder(self, folder_uri: str) -> None:
-        logger.info(self._workspace_folders)
-        del self._workspace_folders[folder_uri]
 
     async def initialize(
-        self, workspaceFolders: Sequence[object], capabilities: Dict[str, object], rootPath=None, rootUri=None, **kwargs
+        self,
+        workspaceFolders: Sequence[lsp_types.WorkspaceFolder],
+        capabilities: Dict[str, object],
+        rootPath=None,
+        rootUri=None,
+        **kwargs,
     ):  # noqa: N803
-        logger.debug("Init: " + json.dumps(kwargs))
-        logger.info(f"workspaceFolders={workspaceFolders}")
+        logger.debug("Init: %s", json.dumps(kwargs))
+        logger.debug("workspaceFolders=%s", workspaceFolders)
+        logger.debug("rootPath=%s", rootPath)
+        logger.debug("rootUri=%s", rootUri)
 
         if rootPath:
             logger.warning("The rootPath parameter has been deprecated in favour of the 'workspaceFolders' parameter.")
         if rootUri:
             logger.warning("The rootUri parameter has been deprecated in favour of the 'workspaceFolders' parameter.")
 
         if workspaceFolders is None:
-            raise InvalidExtensionSetup("No workspace folder specified.")
-
-        # Keep track of the folders opened in this workspace
-        self._workspace_folders: Dict[str, Folder] = Folder.unpack_workspaces(workspaceFolders, self)
-
-        if len(self._workspace_folders) > 1:
-            raise NotImplementedError("Working on multiple folders inside a workspace is not supported yet.")
+            if rootUri is None:
+                raise InvalidExtensionSetup("No workspace folder or rootUri specified.")
+            workspaceFolders = [rootUri]
+        if len(workspaceFolders) > 1:
+            raise InvalidExtensionSetup(
+                "InmantaLSHandler can only handle a single folder. Instantiate one InmantaLSHandler per folder instead."
+            )
 
         init_options = kwargs.get("initializationOptions", None)
 
         if init_options:
-            self.add_setting("inmanta.compiler_venv", init_options.get("compilerVenv", None))
-            self.add_setting("inmanta.repos", init_options.get("repos", None))
+            self.compiler_venv_path = init_options.get(
+                "compilerVenv", os.path.join(os.path.abspath(urlparse(workspaceFolders[0]["uri"]).path), ".env-ls-compiler")
+            )
+            self.repos = init_options.get("repos", None)
+            logger.debug("self.repos= %s", self.repos)
 
+        # Keep track of the root folder opened in this workspace
+        self.root_folder: Folder = Folder(workspaceFolders[0], self)
```

Comment:
You defined the pydantic type but you're not using it. Ideally you'd do the validation a bit further up, so that you can also use the validated object for the compiler venv.